Split by lumi (by work) with event awareness

ericvaandering opened the pull request:
So this now works including the unit tests which come from EventAwareLumiBased.

I changed some numbers because I think I'm actually doing something a little different in that I'm trying to get close the request # of events/job, not stay under it. In other words if I have a 100 event lumi and the job already has 420 events and the target is 500 events, it will add it since 520 is closer to 500 than 400 is.

Otherwise this mostly behaves the same except the correction for split lumis is built in since it constructs jobs by lumi instead of file. The `algorithm()` is about half the size of the EventAwareLumiBased since I'm using LumiList objects and did some other refactoring.

ericvaandering committed:
Split by lumis while paying attention to the amount of work.
This does the final splitting on a lumi by lumi basis, uses LumiList where possible, and naturally doesn't split lumis amoung jobs (no correction function needed).
Also comes with tests.
Refactored several times to make the main loop as compact and clear as possible.

ericvaandering commented:
@ticoann @amaltaro @hufnagel am interested in your thoughts.

@mmascher you might be interested too.

ericvaandering commented:
test this please

ericvaandering commented:
test this please

ericvaandering commented:
retest this please

hufnagel commented on src/python/WMCore/JobSplitting/EventAwareLumiByWork.py:
```diff
@@ -0,0 +1,276 @@
+
+        # First we need to load the data
+        if self.package == 'WMCore.WMBS':
+            loadRunLumi = self.daoFactory(classname="Files.GetBulkRunLumi")
```
This doesn't seem to be used anywhere...

ericvaandering replied:
What do you mean? That the code I have doesn't do anything with it or that this is actually not needed because we don't use this mode of operations? I didn't investigate it, this is just a snippet of code left over from EventAwareByLumi.

hufnagel commented on src/python/WMCore/JobSplitting/EventAwareLumiByWork.py:
```diff
@@ -0,0 +1,276 @@
+            lumisByLocation[locationSet] = LumiList(runsAndLumis=runsAndLumis)
+            if lumiMask:
+                lumisByLocation[locationSet] &= lumiMask
+            lumisToProcess += lumisByLocation[locationSet]
```
Wouldn't it be better to apply the mask earlier, before you build up the full filesByLumi and eventsByLumi dictionaries. If you do sparse processing of large filesets this can make quite a difference...

ericvaandering replied:
I can look into that. Good point.

hufnagel commented:
@ericvaandering , so this basically resorts your input so the splitting algorithm operates on lumis in the input instead of files ? Not sure how useful this is, since the problem is that we want jobs to sequentially read from input files. We do not reorder lumis.

For example, how will this work in the presence of completely unordered input ?

File A : lumi 1,3,5,7,9
File B : lumi 2,4,6,8,10

Each lumi is 1000 events, splitting is set to 2000 events. What kind of jobs is this going to create ?

5 jobs each reading one lumi from file A and another from file B ?

hufnagel commented:
Btw, since our standard WMAgent merge logic doesn't care about lumi order and merges whatever comes first in whatever order it finds, this example case is actually more the norm than the exception.

ericvaandering commented:
Thanks for looking at it Dirk.

In the pathological case you outline, yeah, that's what it would do. Basically when I looked at this I had three goals:

1) Operate on work not files
2) Get rid of one-off code mucking around with lumi lists and move to DataStructs/LumiList
3) Get rid of the post-hoc correction for lumis split across files and build that functionality in from the start.

Ordering was a by product and I think I could rewrite this without much trouble to preserve file order without much trouble.

Re: your 2nd comment, it seems to me (I have not looked at the details yet) that to reduce the # of ACDCs and help with recovery workflows, it would be useful if each step, including merges, could have it's own lumi white list. That way if one output stream is missing set A and another is missing set B, you can re-run the workflow with the union of the two and then automatically filter out what you don't need.

hufnagel commented:
My problem is that this isn't a pathological case, this is the normal case. If you have workflows that run at multiple sites and job assignment is random and then merges happen whenever jobs are done and only at sites, you end up with highly discontinuous assignments of lumis to files.

hufnagel commented:
This job splitter will work ok for workflows that split at single lumi level. At the multiple lumi level the practical effect will be that you end up with jobs that open/read multiple input files and only use a fraction of each file.

hufnagel commented:
You basically will increase the number of file opens across the whole system by a factor that is almost the same as the number of lumis a job is processing on average.

ericvaandering commented:
test this please

ericvaandering commented:
@hufnagel care to take another look? I refactored this a bit to work on files first, addressing your concerns. It's functional as a job splitter, but not on the ACDC stuff yet.

hufnagel commented on src/python/WMCore/JobSplitting/EventAwareLumiByWork.py:
```diff
@@ -0,0 +1,289 @@
+                    continue  # There are no lumis of interest in the file
+                for run, lumi in lumisByFile[lfn].getLumis():
+                    if (run, lumi) in lumisProcessed:
+                        continue  # We already saw this lumi and it got included in an earlier job
```
Is this really needed ? You populate this in the same loop, so this skips if you iterate over the same element twice ? Isn't that impossible ?
